### ISSUE TITLE
"query param" parameters dont re-enter state

### ISSUE THREAD

**Issue by NickFranceschina** [ghempton/ember-routemanager#7]
maybe I'm not understanding how I _should_ be doing this... but I'm trying to use history links like this:

```
http://localhost/#some/path/?sort=name
```

the "query param" part does come thru on stateManager.params... but if I change only the queryparam part

```
http://localhost/#some/path/?sort=date
```

it doesn't seem to call my enter() method again...  not sure what I'm missing

**Comment by ghempton** [comment]
This is definitely a bug.

If you change the params inside the path portion of the route (eg /project/:id/show) it should update correctly.

**Comment by NickFranceschina** [comment]
yes, it does update correctly in that instance.  I've been trying to find the issue, but haven't been able to grok all the code yet :)   the trigger happens, and it gets down into _findState/_matchState... but then I get lost as to what is supposed to be happening

**Comment by NickFranceschina** [comment]
my naive solution... added this line right before the return in _matchState

```
//if old params and new params differ, then "query" was probably appended... also: only dirty if this is a LEAF state
dirty = dirty || state.states && Em.keys(state.states).length == 0 && !Em.isEqual(params, this.get('params'));
```

but probably want to be able to let each state determine if they want to be notified of queryparam changes... since a layout state MIGHT want to know about the queryparams

**Comment by ghempton** [comment]
Yeah this is actually kind of tricky since we do not know what state's use the parameters. It is a lot easier with path parameters because we explicitly know which states should be re-entered.

One solution would be to re-enter _all_ the states, but that seems pretty sub-optimal.

Another would be (as you say) of letting the states indicate that they depend on the query parameters.

I want to give this some more thought and come up with a good general solution.

**Comment by rlivsey** [comment]
I've just run into this myself, so may as well give my 2¢…

Instead of exiting/entering, a state could register its interest in a query parameter and when the state is entered it gets set on the state. That way you could access the params via the usual `get`/`set` instead of through the manager's `params` object.

When query params change, the active states which have registered their interest just get `set` with their new values.

Then the usual Ember bindings could be used to decide what to do at that point.

I have the following route in my app at the moment:

```
messages: Ember.LayoutState.extend({
  route: "messages",
  controller: MB.messagesController,

  enter: function(manager, transition) {
    var params = manager.get("params")
    this.get("controller").set("query", manager.q)
    this.get("controller").fetch()
    this._super(manager, transition)
  }
})
```

The query parameter is used to filter the data in the list, if it changes we don't need to re-render the whole view we just want to notify the controller of the change.

If there was some way of observing the params, then we could do something like:

```
messages: Ember.LayoutState.extend({
  route: "messages",
  queryParams: ["q"],
  controller: MB.messagesController,

  searchQueryChanged: function() {
    this.get("controller").set("query", this.get("q"))
    this.get("controller").fetch()
  }.observes("q")
})
```
